Add PromQL `info` function blog post (#2777)
prometheus/docs: blog/posts/2025-11-14-introducing-info-function.md

---
title: Introducing the Experimental info() Function
created_at: 2025-11-14
kind: article
author_name: Arve Knudsen
---

Enriching metrics with metadata labels can be surprisingly tricky in Prometheus, even if you're a PromQL wiz!
Traditionally, complex PromQL join syntax is required in Prometheus to add even basic information like Kubernetes cluster names or cloud provider regions to queries.
The new, still experimental `info()` function, promises a simpler way, making label enrichment as simple as wrapping your query in a single function call.

In Prometheus 3.0, we introduced the [`info()`](https://prometheus.io/docs/prometheus/latest/querying/functions/#info) function, a powerful new way to enrich your time series with labels from info metrics.
`info` doesn't only offer a simpler syntax however.
It also solves a subtle yet critical problem that has plagued join queries for years: The "churn problem" that causes queries to fail when "non-identifying" info metric labels change.

> In practice, "identifying labels" refers to those labels that the join is performed on.

Author: Do you want to write a section and make a PR to my branch or something? Maybe the easiest.

> In practice, "identifying labels" refers to those labels that the join is performed on.

Reviewer: yeah, can do

Imagine you're monitoring HTTP request durations via OpenTelemetry and want to break them down by Kubernetes cluster.
Your metrics have `job` and `instance` labels, but the cluster name lives in a separate `target_info` metric, as the `k8s_cluster_name` label.
Here's what the traditional approach looks like:

```promql
sum by (k8s_cluster_name, http_status_code) (
    rate(http_server_request_duration_seconds_count[2m])
  * on (job, instance) group_left (k8s_cluster_name)
    target_info
)
```

While this works, there are several issues:

**1. Complexity:** You need to know:
- Which info metric contains your labels (`target_info`)
- Which labels are the "identifying" labels to join on (`job`, `instance`)
- Which data labels you want to add (`k8s_cluster_name`)
- The proper PromQL join syntax (`on`, `group_left`)

This requires expert-level PromQL knowledge and makes queries harder to read and maintain.

**2. The Churn Problem (The Critical Issue):**

Here's the subtle but serious problem: What happens when a Kubernetes pod gets recreated?
The `k8s_pod_name` label in `target_info` changes, and Prometheus sees this as a completely new time series.

> If the old `target_info` series isn't properly marked as stale immediately, both the old and new series can exist simultaneously for up to 5 minutes (the default lookback delta).

Author: Revised, PTAL.

During this overlap period, your join query finds **two distinct matching `target_info` time series** and fails with a "many-to-many matching" error.

This could in practice mean your dashboards break and your alerts stop firing when infrastructure changes are happening, perhaps precisely when you would need visibility the most.

## The Solution: Simple, Reliable Label Enrichment

The `info()` function solves both problems at once.
Here's the same query using `info()`:

```promql
sum by (k8s_cluster_name, http_status_code) (
  info(
    rate(http_server_request_duration_seconds_count[2m]),

>     {k8s_cluster_name=~".+"}

Reviewer 2: I'm wondering if we should better start with the easiest use case first, as this is about showcasing the simplicity. So let's leave out this line (it is also the one "weird" part of the info function, a new type of function parameter, using selector syntax without being a real selector).

It will promote all data labels, but that's not a problem because we aggregate them away in the `sum` anyway. (You can explain that later.)

  )

> )

Author: Revised, PTAL.

```

Much more comprehensible, no?
The real magic happens under the hood though: **`info()` automatically selects the time series with the latest sample**, eliminating churn-related join failures entirely.

> ### Basic Syntax

Reviewer: I know Patrick ran into issues where `info()` had to be put in very specific places, especially when a rate function was involved.  I think his intuition was to "wrap the metric name in info()" but that doesn't work when there's a rate function?  So I think we should have an aside with some advice on how to know where the info function will need to go.

```promql
info(v instant-vector, [data-label-selector instant-vector])
```

- **`v`**: The instant vector to enrich with metadata labels
- **`data-label-selector`** (optional): Label matchers in curly braces to filter which labels to include

If you omit the second parameter, `info()` adds **all** data labels from `target_info`:

```promql
info(rate(http_server_request_duration_seconds_count[2m]))

> ```

Reviewer 2: As said, I would frame this is the primary mode of using `info`. See above.

### Selecting Different Info Metrics

By default, `info()` uses the `target_info` metric.
However, you can select different info metrics (like `build_info`, `node_uname_info`, or `kube_pod_labels`) by including a `__name__` matcher in the data-label-selector:

```promql
# Use build_info instead of target_info
info(up, {__name__="build_info"})

# Use multiple info metrics (combines labels from both)
info(up, {__name__=~"(target|build)_info"})

# Select build_info and only include the version label
info(up, {__name__="build_info", version=~".+"})
```

> **Note:** The current implementation always uses `job` and `instance` as the identifying labels for joining, regardless of which info metric you select.

Author: > "current implementation" implies that this may change, do we have any plans to?

Yes the plan is relatively clear, the last sentence of the same paragraph already explains how it's supposed to work in the future. I don't think we need to solicit feedback, since we know how it's supposed to work on top of persisted metadata.

> This works well for most standard info metrics but may have limitations with custom info metrics that use different identifying labels.

Reviewer 2: Very infamously this would be `kube_pod_labels` from kube-state-metrics. It requires joining on `(namespace, pod)`. This is a pretty common use case. We should not pretend it's fringe and rarely happening, but call out that this doesn't work now but will be fixed in the future (by letting PromQL "know" what labels are identifying).

Author: Revised, PTAL.

## Real-World Use Cases

### OpenTelemetry Integration

The primary driver for the `info()` function is [OpenTelemetry](https://prometheus.io/blog/2024/03/14/commitment-to-opentelemetry/) (OTel) integration.
When using Prometheus as an OTel backend, resource attributes (metadata about the metrics producer) are automatically converted to the `target_info` metric:

- `service.instance.id` → `instance` label
- `service.name` → `job` label
- `service.namespace` → prefixed to `job` (i.e., `<namespace>/<service.name>`)
- All other resource attributes → data labels on `target_info`

This means that, so long as at least either the `service.instance.id` or the `service.name` resource attribute is included, every OTel metric you send to Prometheus over OTLP can be enriched with resource attributes using `info()`:

```promql
# Add all OTel resource attributes
info(rate(http_server_request_duration_seconds_sum[5m]))

# Add only specific attributes
info(
  rate(http_server_request_duration_seconds_sum[5m]),
  {k8s_cluster_name=~".+", k8s_namespace_name=~".+", k8s_pod_name=~".+"}
)

> ```

Reviewer 2: And then add an example that filters for specific values of resource attributes.

### Kubernetes Metadata

Enrich your metrics with Kubernetes-specific information:

```promql
# Add cluster and namespace information to request rates
info(
  sum by (job, http_status_code) (
    rate(http_server_request_duration_seconds_count[2m])
  ),
  {k8s_cluster_name=~".+", k8s_namespace_name=~".+"}

> )

Author: Revised, PTAL.

```

### Cloud Provider Metadata

Add cloud provider information to understand costs and performance by region:

```promql
# Enrich with AWS/GCP/Azure region and availability zone
info(
  rate(cloud_storage_request_count[5m]),
  {cloud_provider=~".+", cloud_region=~".+", cloud_availability_zone=~".+"}
)
```

## Before and After: Side-by-Side Comparison

Let's see how the `info()` function simplifies real queries:

> ### Example 1: Basic Label Enrichment

Reviewer 2: This should include an example where you want to use a lot of resource attributes (because that's very long in the join query, but you just use the default behavior of `info`). Just promoting a single RA is kind-of the worst case for `info` and the best case for the traditional join query, so we shouldn't list it is the 1st example.

  info(
    rate(http_server_request_duration_seconds_count[2m]),
    {k8s_cluster_name=~".+", k8s_namespace_name=~".+"}
  )
)
```

The intent is much clearer with `info`: We're enriching `http_server_request_duration_seconds_count` with cluster and namespace information, then aggregating by those labels and `http_status_code`.

## Technical Benefits

Beyond cleaner syntax, the `info()` function provides several technical advantages:

### 1. Automatic Churn Handling

As previously mentioned, `info()` automatically picks the matching info time series with the latest sample when multiple versions exist.
This eliminates the "many-to-many matching" errors that plague traditional join queries during churn.

**How it works:** When non-identifying info metric labels change (e.g., a pod is re-created), there's a brief period where both old and new series might exist.
The `info()` function simply selects whichever has the most recent sample, ensuring your queries keep working.

### 2. Better Performance

The `info()` function is more efficient than traditional joins:
- Only selects matching info series
- Avoids unnecessary label matching operations

> - Optimized query execution path

Author: I don't know about perf comparisons versus label promotion. I just know there's an obvious performance benefit over join queries, due to being able to filter what to select on the RHS :)

## Getting Started

The `info()` function is experimental and must be enabled via a feature flag:

```bash
prometheus --enable-feature=promql-experimental-functions
```

Once enabled, you can start using it immediately.
Here are some simple examples to try:

```promql
# Basic usage - add all target_info labels
info(up)

# Selective enrichment - add only cluster name
info(up, {k8s_cluster_name=~".+"})

# In a real query
info(
  rate(http_server_request_duration_seconds_count[5m]),
  {k8s_cluster_name=~".+"}
)

# With aggregation
sum by (k8s_cluster_name) (
  info(up, {k8s_cluster_name=~".+"})
)

> ```

Reviewer 2: Not sure if there is value to repeat examples already given above.

## Current Limitations and Future Plans

The current implementation is an **MVP (Minimum Viable Product)** designed to validate the approach and gather user feedback.
It has some intentional limitations:

### Current Constraints

1. **Default info metric:** Only considers `target_info` by default
   - Workaround: You can use `__name__` matchers like `{__name__=~"(target|build)_info"}` in the data-label-selector, though this still assumes `job` and `instance` as identifying labels

2. **Fixed identifying labels:** Always assumes `job` and `instance` are the identifying labels for joining
   - This works for most use cases but may not be suitable for all scenarios

### Future Development

These limitations are meant to be temporary.
The experimental status allows us to:
- Gather real-world usage feedback
- Understand which use cases matter the most
- Iterate on the design before committing to a final API

A future version of the `info()` function should:
- Support all info metrics (not just `target_info`)

> - Dynamically determine identifying labels based on the info metric's structure

Reviewer 2: It's not so much dynamic, but it's about (statically) storing the information what labels are identifying.

Note for future implementations (not related to the review, just happened to come to my mind while doing this review): For something like `kube_pod_labels`, we'll have the actual identifying labels (`namespace`, `pod`), then we'll have the data labels (in this case the actual pod labels), but when these metrics are ingested from KSM, we will also get a `job` and `instance` label attached (and whatever other target labels are configured). The future perfect version of `info` probably should just join on `namespace` and `pod` and only add the pod labels, but completely ignore `job` and `instance` and possibly other target labels (i.e. don't join on them, but also don't add them to the result either). Devil is in the detail here…

- Which data labels you want to add (`k8s_cluster_name`)
- The proper PromQL join syntax (`on`, `group_left`)

This requires expert-level PromQL knowledge and makes queries harder to read and maintain.

**2. The Churn Problem (The Critical Issue):**

Here's the subtle but serious problem: What happens when an OTel resource attribute changes in a Kubernetes container, while the identifying resource attributes stay the same?
An example could be the resource attribute `k8s.pod.labels.app.kubernetes.io/version`.
Then the corresponding `target_info` label `k8s.pod.labels.app.kubernetes.io/version` changes, and Prometheus sees a completely new `target_info` time series.

As the OTLP endpoint doesn't mark the old `target_info` series as stale, both the old and new series can exist simultaneously for up to 5 minutes (the default lookback delta).
During this overlap period, your join query finds **two distinct matching `target_info` time series** and fails with a "many-to-many matching" error.

This could in practice mean your dashboards break and your alerts stop firing when infrastructure changes are happening, perhaps precisely when you would need visibility the most.

> 

Author: You're right. On reviewing the section, I realized that a sub-section was missing. I added it, PTAL.

```promql
sum by (k8s_cluster_name, http_status_code) (
  info(rate(http_server_request_duration_seconds_count[2m]))
)
```

> Much more comprehensible, no?

Reviewer 2: Language nit:

```suggestion
Much more comprehensible, isn't it?
```

Unless you intentionally want to use a specific sociolect that is not standard English…

```promql
# Add all OTel resource attributes
info(rate(http_server_request_duration_seconds_sum[5m]))

# Add only specific attributes
info(
  rate(http_server_request_duration_seconds_sum[5m]),
  {k8s_cluster_name=~".+", k8s_namespace_name=~".+", k8s_pod_name=~".+"}
)
```

### Build Information

Enrich your metrics with build-time information:

```promql
# Add version and branch information to request rates
sum by (job, http_status_code, version, branch) info(
  rate(http_server_request_duration_seconds_count[2m]),
  {__name__="build_info"}
)
```

### Filter on Producer Version

Pick only metrics from certain producer versions:

```promql
sum by (job, http_status_code, version) info(
  rate(http_server_request_duration_seconds_count[2m]),
  {__name__="build_info", version=~"2\..+"}
)
```

## Before and After: Side-by-Side Comparison

Let's see how the `info()` function simplifies real queries:

### Example 1: OpenTelemetry Resource Attribute Enrichment

**Traditional approach:**
```promql

> sum by (http_status_code, k8s_cluster_name, k8s_container_name, k8s_cronjob_name, k8s_daemonset_name, k8s_deployment_name, k8s_job_name, k8s_namespace_name, k8s_pod_name, k8s_replicaset_name, k8s_statefulset_name) (

Reviewer 2: Is this a realistic set of labels to sum by? I don't know the semantic conversions by heart, but isn't `k8s_pod_name` essentially unique per pod? There wouldn't be much to sum up in this way… Maybe sum by fewer labels, something that is more likely to happen in a real query.

(Mostly I want to prevent that readers just dismiss the query as contrived.)

Author: I revised to what might be a more realistic set of labels. PTAL.

    rate(http_server_request_duration_seconds_count[2m])
  * on (job, instance) group_left (k8s_cluster_name, k8s_container_name, k8s_cronjob_name, k8s_daemonset_name, k8s_deployment_name, k8s_job_name, k8s_namespace_name, k8s_pod_name, k8s_replicaset_name, k8s_statefulset_name)
    target_info
)
```

**With info():**
```promql
sum by (http_status_code, k8s_cluster_name, k8s_container_name, k8s_cronjob_name, k8s_daemonset_name, k8s_deployment_name, k8s_job_name, k8s_namespace_name, k8s_pod_name, k8s_replicaset_name, k8s_statefulset_name) (
  info(rate(http_server_request_duration_seconds_count[2m]))
)
```

The intent is much clearer with `info`: We're enriching `http_server_request_duration_seconds_count` with Kubernetes related OpenTelemetry resource attributes.

### Example 2: Filtering by Label Value

**Traditional approach:**
```promql
sum by (http_status_code) (
    rate(http_server_request_duration_seconds_count[2m])
  * on (job, instance) group_left ()
    target_info{k8s_cluster_name="us-east-1"}
)
```

**With info():**
```promql
sum by (http_status_code) (
  info(
    rate(http_server_request_duration_seconds_count[2m]),
    {k8s_cluster_name="us-east-1"}
  )

> )

Reviewer 2: With above change, this would look like:

```suggestion
sum by (http_status_code, k8s_cluster_name) (
  info(
    rate(http_server_request_duration_seconds_count[2m]),
    {k8s_cluster_name=~"us-.*"}
  )
)
```

```

> Here we filter to only include metrics from the `us-east-1` cluster. The `info()` version integrates the filter naturally into the data-label-selector.

Reviewer 2: ```suggestion
Here we filter to only include metrics from clusters in the US (whose name starts with `us-`). The `info()` version integrates the filter naturally into the data-label-selector.
```
